auth/src/main/java/com/firebase/ui/auth/util/SmartLock.java

```diff
@@ -133,77 +113,149 @@ public void onConnectionFailed(@NonNull ConnectionResult connectionResult) {
-            Context context,
-            FlowParameters flowParams,
-            FirebaseUser user,
-            @Nullable String password,
-            @Nullable String provider) {
-
-        String photoUrl = user.getPhotoUrl() != null ? user.getPhotoUrl().toString() : null;
-        return ActivityHelper.createBaseIntent(context, SaveCredentialsActivity.class, flowParams)
-                .putExtra(ExtraConstants.EXTRA_NAME, user.getDisplayName())
-                .putExtra(ExtraConstants.EXTRA_EMAIL, user.getEmail())
-                .putExtra(ExtraConstants.EXTRA_PASSWORD, password)
-                .putExtra(ExtraConstants.EXTRA_PROVIDER, provider)
-                .putExtra(ExtraConstants.EXTRA_PROFILE_PICTURE_URI, photoUrl);
+    /**
+     * If SmartLock is enabled and Google Play Services is available, save the credentials.
+     * Otherwise, finish the calling Activity with RESULT_OK.
+     *
+     * @param activity     the calling Activity.
+     * @param parameters   calling Activity flow parameters.
+     * @param firebaseUser Firebase user to save in Credential.
+     * @param password     (optional) password for email credential.
+     * @param provider     (optional) provider string for provider credential.
+     */
+    public void saveCredentialsOrFinish(AppCompatBase activity,
+                                        FlowParameters parameters,
+                                        FirebaseUser firebaseUser,
+                                        @Nullable String password,
+                                        @Nullable String provider) {
+        mFlowParameters = parameters;
+        mName = firebaseUser.getDisplayName();
+        mEmail = firebaseUser.getEmail();
+        mPassword = password;
+        mProvider = provider;
+        mProfilePictureUri = firebaseUser.getPhotoUrl() != null ? firebaseUser.getPhotoUrl()
+                .toString() : null;
+
+        // If SmartLock is disabled, finish the Activity
+        if (!parameters.smartLockEnabled) {
+            activity.finish(RESULT_OK, activity.getIntent());
+            return;
+        }
+
+        // If Play Services is not available, finish the Activity
+        if (!PlayServicesHelper.getInstance(activity).isPlayServicesAvailable()) {
+            activity.finish(RESULT_OK, activity.getIntent());
+            return;
+        }
+
+        if (!FirebaseAuthWrapperFactory.getFirebaseAuthWrapper(parameters.appName)
+                .isPlayServicesAvailable(activity)) {
+            activity.finish(RESULT_OK, activity.getIntent());
+            return;
+        }
+
+        mActivity = activity;
+        mCredentialsApiClient = new GoogleApiClient.Builder(activity)
+                .addConnectionCallbacks(this)
+                .addOnConnectionFailedListener(this)
+                .addApi(Auth.CREDENTIALS_API)
+                .enableAutoManage(activity, this)
+                .build();
+    }
+
+    public Intent getIntentForTest() {
```

Comment:
This is returning an intent when the SmartLock doesn't actually use intents.  This makes the tests slightly misleading IMO.  See the comments on the tests below.